### DISCUSSION TITLE
Coupled flux term in dusty gas model

### DISCUSSION THREAD

**Discussion by dfukuyama** [usnistgov/fipy#994] [Q&A] [answered]
I have been trying to implement a porous media dusty gas model (similar to Stefan-Maxwell diffusion) in fipy, and I am curious to know if there is a good way to implement the flux term. The mass conservation equation is as follows:
$$\frac{\partial n_i \phi}{\partial t} + \nabla \cdot J_i = 0$$
In this model, the flux terms are coupled:
$$\frac{J_i}{D_{ik}}+\sum_{i\neq j}\frac{n_j J_i - n_i J_j}{nD_{ij}}=\frac{n_i}{D_{ik}}q-\nabla n_i$$
Here, $\phi$ is the porosity, the $D$ values are the various diffusivity coefficients, $J_{i}$ is the flux term, $n_{i}$ are the molar densities of the gas components, $n$ is the total molar density of the gas, and q is the Darcy flux term:
$$q=-\frac{kRT}{\mu}\nabla P$$

For two components, the matrix-vector form is:
```math
\begin{bmatrix}\frac{D_12}{D_1k}+y_2 & -y_1 \\ -y_2 & \frac{D_{12}}{D_{2k}}+y_1\end{bmatrix}\cdot\begin{bmatrix}J_1\\J_2\end{bmatrix}=D_{12}\begin{bmatrix}\frac{n_1}{D_{1k}}q-\nabla n_1 \\ \frac{n_2}{D_{2k}}q-\nabla n_2\end{bmatrix}
```
where $y$ is the mole fraction, $y_i=\frac{n_i}{n}$.

I am thinking of taking the analytic inverse of the left-hand side "A" matrix, multiplying through, and casting the result into four different ```DiffusionTerm```s for each equation. Is there a better way to deal with this coupled flux term?

**Comment by guyer** [comment, marked as answer]
I hoped there might be something clever to do with taking the divergence of the second equation and then substituting the first equation, but I think your approach is correct. I'm don't think FiPy offers anything to make it easier.

**Reply by dfukuyama** [reply]
Thanks!
